[ONNX] A better way to safe guard 2GB  model serialization

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111984

Summary
- faster than previous try-catch.
- more stable than previous try-catch. In some circumstances serializing models > 2GB into a single protobuf file ends up with a corrupted file without raising an exception.

BowenBao committed:
[ONNX] A better way to safe guard 2GB  model serialization

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -572,15 +572,16 @@ def serialize(
         import onnx
 
-        try:
+        if export_output.model_proto.ByteSize() < 1.9 * 1024 * 1024 * 1024:
```
Make the threshold a constant for readability?

thiagocrepaldi reviewed (commented):
Depending on magic numbers is usually not reliable and might need tweaks from time to time.

Ideally, we should find out the root cause on why we get a corrupted protobuf when the try-catch doesn't raise

Is there a repro that motivated this change?

BowenBao committed:
Update on "[ONNX] A better way to safe guard 2GB  model serialization"


Summary
- faster than previous try-catch.
- more stable than previous try-catch. In some circumstances serializing models > 2GB into a single protobuf file ends up with a corrupted file without raising an exception.


[ghstack-poisoned]

BowenBao commented:
This is the most proper way. `try-catch` was a hack because back then I didn't realize we can access the accurate total bytesize. Back in `export.cpp` for torchscript exporter we cannot.

This issue was discovered by models from the torchbench. It may be repro-ed by exporting a large model such as llama2 or stable diffusion xl. However the issue only happened recently. Feel free to take a look if you have bandwidth.

BowenBao commented:
@pytorchbot merge
